packages/labs/forms/src/form-associated.ts

```diff
@@ -0,0 +1,227 @@
+/**
+ * @license
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {type ReactiveElement} from '@lit/reactive-element';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Constructor<T = object> = new (...args: any[]) => T;
+
+/**
+ * Generates a public interface type that removes private and protected fields.
+ * This allows accepting otherwise incompatible versions of the type (e.g. from
+ * multiple copies of the same package in `node_modules`).
+ */
+type Interface<T> = {
+  [K in keyof T]: T[K];
+};
+
+interface FormAssociated extends ReactiveElement {
+  _checkValidity?(): ValidityResult;
+}
+
+interface FormAssociatedConstructor {
+  [Symbol.metadata]: object & Record<PropertyKey, unknown>;
+
+  role: ElementInternals['role'];
+
+  new (): FormAssociated;
+}
+
+interface ValidityResult {
+  flags: ValidityState;
+  message?: string;
+  anchor?: HTMLElement;
+}
+
+const internalsCache = new WeakMap<FormAssociated, ElementInternals>();
+
+/**
+ * Returns the ElementInternals for a FormAssociated element. This should only
+ * be called by subclasses of FormAssociated.
+ */
+export const getInternals = (element: FormAssociated) =>
+  internalsCache.get(element);
+
+/**
+ * Returns true if the element is disabled.
+ */
+export const isDisabled = (element: FormAssociated) =>
+  disabledElements.has(element);
+
+const disabledElements = new WeakSet<FormAssociated>();
+
+/**
+ * A mixin that makes a ReactiveElement into a form-associated custom element.
+ *
+ * - ARIA role: Sets this element's ElementInternals role to the value of the
+ *   static `role` property.
+ * - Form value: The value of the element is stored in the field decorated with
+ *   the `@formValue()` decorator. This value field can have any name, but the
+ *   type must be assignable to `string | File | FormData | null`.
+ * - Form state: The state of the element is stored in the field decorated with
```

Comment:
I don't follow why this exists and is separate from the value. If it's to provide additional control, couldn't this just be done with value + user customization?

Reply:
The browser separates form state and value, in `setFormValue()` (which I need to call w/ state still) and `formStateRestoreCallback()`. This is to be able to save other state that isn't necessarily part of the submitted form value. You could serialize some transient UI state so when the form is restored you restore that UI.

I'm not sure how often it's used. It's weird to me that it has to have the same type as value. I can see how you can encode extra state in a string, but not how you can do that with a File object.

Reply:
ok, yeah. There's an example here: https://web.dev/articles/more-capable-form-controls#restoring-form-state